File: backend/src/tasks/xpanse-sync.ts
import { int } from 'aws-sdk/clients/datapipeline';
import { CommandOptions } from './ecs-client';
import { Cpe, Cve } from '../models';
import axios from 'axios';
import { plainToClass } from 'class-transformer';
import {
  connectToDatabase,
  Domain,
  Service,
  Vulnerability,
  Organization
} from '../models';
import { isIP } from 'net';
import * as dns from 'dns';
import saveDomainsReturn from './helpers/saveDomainsReturn';

interface XpanseVulnOutput {
  alert_id: string;
  alert_name: string;
  description: string;
  last_modified_ts: Date;
  local_insert_ts: Date;
  event_timestamp: Date[];
  host_name: string;
  alert_action: string;
  action_country: string[];
  action_remote_port: number[];
  external_id: string;
  related_external_id: string;
  alert_occurrence: number;
  severity: string;
  matching_status: string;
  alert_type: string;
  resolution_status: string;
  resolution_comment: string;
  last_observed: string;
  country_codes: string[];
  cloud_providers: string[];
  ipv4_addresses: string[];
  domain_names: string[];
  port_protocol: string;
  time_pulled_from_xpanse: string;
  action_pretty: string;
  attack_surface_rule_name: string;
  certificate: Record<string, any>;
  remediation_guidance: string;
  asset_identifiers: Record<string, any>[];
  services: XpanseServiceOutput[];
}
interface XpanseServiceOutput {
  service_id?: string;
  service_name?: string;
  service_type?: string;
  ip_address?: string[];
  domain?: string[];
  externally_detected_providers?: string[];
  is_active?: string;
  first_observed?: string;
  last_observed?: string;
  port?: number;
  protocol?: string;
  active_classifications?: string[];
  inactive_classifications?: string[];
  discovery_type?: string;
  externally_inferred_vulnerability_score?: string;
  externally_inferred_cves?: string[];
  service_key?: string;
  service_key_type?: string;
  cves?: XpanseCveOutput[];
}
interface XpanseCveOutput {
  cve_id?: string;
  cvss_score_v2?: string;
  cve_severity_v2?: string;
  cvss_score_v3?: string;
  cve_severity_v3?: string;
  inferred_cve_match_type?: string;
  product?: string;
  confidence?: string;
  vendor?: string;
  version_number?: string;
  activity_status?: string;
  first_observed?: string;
  last_observed?: string;
}
interface TaskResponse {
  task_id: string;
  status: string;
  result: null | {
    total_pages: number;
    current_page: number;
    data: XpanseVulnOutput[];
  };
  error: string | null;
}
function isIPAddress(input: string): boolean {
  // Regular expression to match an IP address
  const ipRegex = /^(?:\d{1,3}\.){3}\d{1,3}$/;

  return ipRegex.test(input);
}

const createTask = async (page: number, orgId: string) => {
  console.log('Creating task to fetch Xpanse Alert data');
  try {
    const response = await axios({
      url: 'https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/xpanse_vulns',
      method: 'POST',
      headers: {
        Authorization: String(process.env.CF_API_KEY),
        access_token: String(process.env.PE_API_KEY),
        'Content-Type': '' //this is needed or else it breaks because axios defaults to application/json
      },
      data: {
        org_acronym: orgId,
        page: page,
        per_page: 100 // Tested with 150 and 200 but this results in 502 errors on certain pages with a lot of CPEs
      }
    });
    if (response.status >= 200 && response.status < 300) {
      //console.log('Request was successful');
    } else {
      console.log('Request failed');
    }
    return response.data as TaskResponse;
  } catch (error) {
    console.log(`Error making POST request: ${error}`);
  }
};
const fetchData = async (task_id: string) => {
  console.log('Fetching Xpanse alert data');
  try {
    const response = await axios({
      url: `https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/xpanse_vulns/task/${task_id}`,
      headers: {
        Authorization: String(process.env.CF_API_KEY),
        access_token: String(process.env.PE_API_KEY),
        'Content-Type': ''
      }
    });
    if (response.status >= 200 && response.status < 300) {
      //console.log('Request was successful');
    } else {
      console.log('Request failed');
    }
    return response.data as TaskResponse;
  } catch (error) {
    console.log(`Error making POST request: ${error}`);
  }
};
const getVulnData = async (
  org: Organization,
  commandOptions: CommandOptions
) => {
  let done = false;
  let page = 1;
  let total_pages = 1;
  while (!done) {
    let taskRequest = await createTask(page, org.acronym);
    console.log(`Fetching page ${page} of page ${total_pages}`);
    await new Promise((r) => setTimeout(r, 1000));
    if (taskRequest?.status == 'Processing') {
      while (taskRequest?.status == 'Processing') {
        await new Promise((r) => setTimeout(r, 1000));
        taskRequest = await fetchData(taskRequest.task_id);
      }
      if (taskRequest?.status == 'Completed') {
        console.log(`Task completed successfully for page: ${page}`);
        const vulnArray = taskRequest?.result?.data || [];
        await saveXpanseAlert(vulnArray, org, commandOptions.scanId);
        total_pages = taskRequest?.result?.total_pages || 1;
        const current_page = taskRequest?.result?.current_page || 1;
        if (current_page >= total_pages) {
          done = true;
          console.log(`Finished fetching Xpanse Alert data for ${org.name}`);
        }
        page = page + 1;
      }
    } else {
      done = true;
      console.log(
        `Error fetching Xpanse Alert data: ${taskRequest?.error} and status: ${taskRequest?.status}`
      );
    }
  }
  return 1;
};
const saveXpanseDomain = async (domain) => {
  const existingDomain = await Domain.findOne({
    where: {
      name: domain.name,
      organization: { id: domain.organization.id }
    },
    relations: ['organization']
  });
  if (!existingDomain) {
    console.log(`Saving domain ${domain.name}`);
    const newDomain = await Domain.save(domain);
    return newDomain;
  } else {
    return existingDomain;
  }
}; // save vulns if this vulns hasn't been seen or if last_seen >= last_seen on alert in db
const saveXpanseVuln = async (vuln: Vulnerability, savedDomain: Domain) => {
  const vulns: Vulnerability[] = [];
  const existingVuln = await Vulnerability.findOne({
    where: {
      domain: { id: savedDomain.id },
      title: vuln.title
    }
  });
  if (!existingVuln) {
    const newVuln = await Vulnerability.save(vuln);
    return newVuln;
  } else {
    if (vuln.lastSeen) {
      if (existingVuln.lastSeen) {
        if (existingVuln.lastSeen >= vuln.lastSeen) {
          return existingVuln;
        }
      }
      existingVuln.lastSeen = new Date(vuln.lastSeen);
    }
    existingVuln.state = vuln.state;
    existingVuln.structuredData = vuln.structuredData;
    existingVuln.domain = savedDomain;
    const newVuln = await Vulnerability.save(existingVuln);
    return newVuln;
  }
};
const saveXpanseAlert = async (
  alerts: XpanseVulnOutput[],
  org: Organization,
  scan_id: string
) => {
  for (const vuln of alerts) {
    let savedDomain: Domain;
    let service_domain;
    let service_ip;
    let ipOnly = false;
    try {
      if (isIPAddress(vuln.host_name)) {
        service_ip = vuln.host_name;
        try {
          service_domain = (await dns.promises.reverse(service_ip))[0];
        } catch {
          service_domain = service_ip;
          ipOnly = true;
        }
        service_ip = vuln.host_name;
      } else {
        service_domain = vuln.host_name;
        try {
          service_ip = (await dns.promises.lookup(service_domain)).address;
        } catch {
          service_ip = null;
        }
      }
      savedDomain = await saveXpanseDomain(
        plainToClass(Domain, {
          name: service_domain.toLowerCase(),
          ip: service_ip,
          organization: org,
          fromRootDomain: !ipOnly
            ? service_domain.split('.').slice(-2).join('.')
            : null,
          discoveredBy: { id: scan_id },
          subdomainSource: `Palo Alto Xpanse`,
          ipOnly: ipOnly
        })
      );
    } catch (e) {
      console.error(`Failed to save domain for host: ${vuln.host_name}`);
      console.error(e);
      console.error('Continuing to next vulnerability');
      continue;
    }

    try {
      let resolution;
      if (vuln.resolution_status.includes('RESOLVED')) {
        resolution = 'closed';
      } else {
        resolution = 'open';
      }
      const vuln_obj: Vulnerability = plainToClass(Vulnerability, {
        domain: savedDomain,
        lastSeen: vuln.last_modified_ts,
        title: vuln.alert_name,
        cve: 'Xpanse Alert',
        description: vuln.description,
        severity:
          vuln.severity !== null
            ? vuln.severity.charAt(0).toUpperCase() +
              vuln.severity.slice(1).toLowerCase()
            : null,
        state: resolution,
        structuredData: {
          alert_id: vuln.alert_id,
          xpanse_modified_ts: vuln.last_modified_ts,
          xpanse_insert_ts: vuln.local_insert_ts,
          xpanse_last_observed: vuln.last_observed,
          event_timestamps: vuln.event_timestamp,
          host_name: vuln.host_name,
          action_countries: vuln.action_country,
          action_remote_port: vuln.action_remote_port,
          external_id: vuln.external_id,
          related_external_id: vuln.related_external_id,
          alert_occurrence: vuln.alert_occurrence,
          resolution_status: vuln.resolution_status,
          resolution_comment: vuln.resolution_comment,
          country_codes: vuln.country_codes,
          cloud_providers: vuln.cloud_providers,
          ipv4_addresses: vuln.ipv4_addresses,
          domain_names: vuln.domain_names,
          port_protocol: vuln.port_protocol,
          time_pulled_from_xpanse: vuln.time_pulled_from_xpanse,
          attack_surface_rule_name: vuln.attack_surface_rule_name,
          certificate: vuln.certificate,
          remediation_guidance: vuln.remediation_guidance,
          asset_identifiers: vuln.asset_identifiers
        },
        source: `Palo Alto Xpanse`,
        needsPopulation: true,
        service: null
      });
      await saveXpanseVuln(vuln_obj, savedDomain);
      // save vulns if this vulns hasn't been seen or if last_seen >= last_seen on alert in db
    } catch (e) {
      console.error('Could not save vulnerability. Continuing.');
      console.error(e);
    }
  }
};

export const handler = async (CommandOptions) => {
  try {
    await connectToDatabase();
    const allOrgs: Organization[] = await Organization.find();

    for (const org of allOrgs) {
      console.log(`Gathering Xpanse alerts for ${org.name}`);
      await getVulnData(org, CommandOptions);
    }
  } catch (e) {
    console.error('Unknown failure.');
    console.error(e);
  }
};

File: backend/src/xfd_django/xfd_api/tasks/xpanse_sync.py
"""XpanseSync scan."""
# Standard Python Libraries
import datetime
import logging
import os
import random
import time

# Third-Party Libraries
import django
from django.conf import settings
from django.core.exceptions import ObjectDoesNotExist
import requests
from xfd_api.helpers.date_time_helpers import calculate_days_back
from xfd_mini_dl.models import (
    Organization,
    XpanseAlerts,
    XpanseBusinessUnits,
    XpanseCveServiceMdl,
    XpanseCvesMdl,
    XpanseServicesMdl,
)

# Django setup
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "xfd_django.settings")
os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true"
django.setup()

LOGGER = logging.getLogger(__name__)

# Constants
MAX_RETRIES = 3  # Max retries for failed tasks

headers = settings.DMZ_API_HEADER


def handler(event):
    """Retrieve and save Xpanse alerts from the DMZ."""
    try:
        main()
        return {
            "status_code": 200,
            "body": "DMZ Xpanse Assets sync completed successfully.",
        }
    except Exception as e:
        return {"status_code": 500, "body": str(e)}


def main():
    """Fetch and save DMZ Xpanse alerts."""
    try:
        modified_timestamp_str = calculate_days_back(5)

        if is_bu_pull_day() or XpanseBusinessUnits.objects.count() == 0:
            business_units = pull_and_save_business_units()
        else:
            business_units = list(XpanseBusinessUnits.objects.all())

        random.shuffle(business_units)

        for business_unit in business_units:
            done = False
            page = 1
            total_pages = 2
            per_page = 100
            retry_count = 0
            if business_unit.cyhy_db_name:
                while not done:
                    data = fetch_dmz_xpanse_alert_task(
                        business_unit.cyhy_db_name.acronym,
                        page,
                        per_page,
                        modified_timestamp_str,
                    )
                    if not data or data.get("status") != "Processing":
                        retry_count += 1

                        if retry_count >= MAX_RETRIES:
                            LOGGER.warning(
                                "Max retries reached for org: %s. Moving to next organization.",
                                business_unit.cyhy_db_name.acronym,
                            )
                            break  # Skip to next organization

                        time.sleep(5)
                        continue
                    response = fetch_dmz_xpanse_data(data.get("task_id"))

                    while response and response.get("status") == "Pending":
                        time.sleep(1)
                        response = fetch_dmz_xpanse_data(data.get("task_id", None))
                    if response and response.get("status") == "Completed":
                        xpanse_alerts = response.get("result", {}).get("data", {})
                        total_pages = response.get("result", {}).get("total_pages", 1)
                        current_page = response.get("result", {}).get("current_page", 1)
                        save_alerts_to_db(xpanse_alerts)
                        LOGGER.info(len(xpanse_alerts))
                        if current_page >= total_pages:
                            done = True
                        page += 1
                    else:
                        LOGGER.exception(
                            "Task error: %s - Status: %s",
                            response.get("error"),
                            response.get("status"),
                        )
                        raise Exception
            else:
                LOGGER.warning(
                    "%s does not have a linked CyHy org", business_unit.entity_name
                )
    except Exception as e:
        LOGGER.error("Scan failed to complete: %s", e)


def is_bu_pull_day():
    """Check if today is a day to repull all business units."""
    today = datetime.datetime.today()
    day_of_month = today.day
    return day_of_month in (17, 2)


def pull_and_save_business_units():
    """Pull xpanse business units and save to db."""
    LOGGER.info("Fetching Xpanse Business Units")
    headers = {
        "X-API-KEY": os.getenv("CF_API_KEY"),
        "access_token": os.getenv("PE_API_KEY"),
        "Content-Type": "",
    }

    try:
        response = requests.get(
            "https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/linked_mdl_xpanse_business_units",
            headers=headers,
            timeout=20,  # Timeout in seconds
        )
        response.raise_for_status()

    except requests.exceptions.RequestException as e:
        LOGGER.error("Error fetching DMZ Business Unit pull: %s", e)
        return None
    try:
        bu_list = []
        business_unit_list = response.json()

        for business_unit in business_unit_list:
            if business_unit.get("cyhy_db_name"):
                try:
                    organization = Organization.objects.get(
                        acronym=business_unit.get("cyhy_db_name")
                    )
                except ObjectDoesNotExist:
                    organization = None

                mdl_defaults = {
                    "state": business_unit.get("state"),
                    "county": business_unit.get("county"),
                    "city": business_unit.get("city"),
                    "sector": business_unit.get("sector"),
                    "entity_type": business_unit.get("entity_type"),
                    "region": business_unit.get("region"),
                    "rating": business_unit.get("rating"),
                    "cyhy_db_name": organization,
                }

                (
                    mdl_business_unit_object,
                    mdl_created,
                ) = XpanseBusinessUnits.objects.update_or_create(
                    entity_name=business_unit.get("entity_name"), defaults=mdl_defaults
                )
                bu_list.append(mdl_business_unit_object)

        LOGGER.info("Business Units saved to MDL.")
        return bu_list
    except Exception as e:
        LOGGER.error("Error fetching DMZ Business Unit pull: %s", e)
        return None


def fetch_dmz_xpanse_alert_task(org_acronym, page, per_page, modified_datetime):
    """Fetch xpanse alert task id."""
    LOGGER.info("Fetching xpanse alert task for organization: %s", org_acronym)
    headers = {
        "X-API-KEY": os.getenv("CF_API_KEY"),
        "access_token": os.getenv("PE_API_KEY"),
        "Content-Type": "",
    }

    data = {
        "org_acronym": org_acronym,
        "page": page,
        "per_page": per_page,
        "modified_datetime": modified_datetime,
    }

    try:
        response = requests.post(
            "https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/get_mdl_xpanse_alert",
            headers=headers,
            json=data,
            timeout=20,  # Timeout in seconds
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        LOGGER.error("Error fetching DMZ task: %s", e)
        return None


def fetch_dmz_xpanse_data(task_id):
    """Fetch DMZ xpanse alert data for a task."""
    url = "https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/mdl_xpanse_alerts_task_status/task/{task_id}".format(
        task_id=task_id
    )
    headers = {
        "X-API-KEY": os.getenv("CF_API_KEY"),
        "access_token": os.getenv("PE_API_KEY"),
        "Content-Type": "",
    }

    try:
        response = requests.get(url, headers=headers, timeout=20)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        LOGGER.error("Error fetching DMZ Xpanse data: %s", e)
        return None


def save_alerts_to_db(alert_list):
    """Save a list of Xpanse alerts to the MDL."""
    for alert in alert_list:
        try:
            defaults = {
                "time_pulled_from_xpanse": alert.get("time_pulled_from_xpanse", None),
                "detection_timestamp": alert.get("detection_timestamp", None),
                "alert_name": alert.get("alert_name", None),
                "description": alert.get("description", None),
                "host_name": alert.get("host_name", None),
                "alert_action": alert.get("alert_action", None),
                "action_pretty": alert.get("action_pretty", None),
                "action_country": alert.get("action_country", None),
                "action_remote_port": alert.get("action_remote_port", None),
                "starred": alert.get("starred", None),
                "external_id": alert.get("external_id", None),
                "related_external_id": alert.get("related_external_id", None),
                "alert_occurrence": alert.get("alert_occurrence", None),
                "severity": alert.get("severity", None),
                "matching_status": alert.get("matching_status", None),
                "local_insert_ts": alert.get("local_insert_ts", None),
                "last_modified_ts": alert.get("last_modified_ts", None),
                "case_id": alert.get("case_id", None),
                "event_timestamp": alert.get("event_timestamp", None),
                "alert_type": alert.get("alert_type", None),
                "resolution_status": alert.get("resolution_status", None),
                "resolution_comment": alert.get("resolution_comment", None),
                "tags": alert.get("tags", None),
                "last_observed": alert.get("last_observed", None),
                "country_codes": alert.get("country_codes", None),
                "cloud_providers": alert.get("cloud_providers", None),
                "ipv4_addresses": alert.get("ipv4_addresses", None),
                "domain_names": alert.get("domain_names", None),
                "service_ids": alert.get("service_ids", None),
                "website_ids": alert.get("website_ids", None),
                "asset_ids": alert.get("asset_ids", None),
                "certificate": alert.get("certificate", None),
                "port_protocol": alert.get("port_protocol", None),
                "attack_surface_rule_name": alert.get("attack_surface_rule_name", None),
                "remediation_guidance": alert.get("remediation_guidance", None),
                "asset_identifiers": alert.get("asset_identifiers", None),
            }

            alert_object, created = XpanseAlerts.objects.update_or_create(
                alert_id=alert.get("alert_id"),
                defaults=defaults,
            )

            business_unit_list = []
            tags = (alert.get("tags", None),)
            for tag in tags[0]:
                if tag.startswith("BU:"):
                    business_unit_list.append(tag[3:].strip())

            business_unit_objects = []
            for b_u in business_unit_list:
                try:
                    business_unit_objects.append(
                        XpanseBusinessUnits.objects.get(entity_name=b_u)
                    )
                except Exception as e:
                    LOGGER.error("Failed to get business unit %s: %s", b_u, e)
                    continue

            alert_object.business_units.set(business_unit_objects)

            service_objects = []
            for service in alert.get("services", []):
                service_defaults = {
                    "service_name": service.get("service_name"),
                    "service_type": service.get("service_type"),
                    "ip_address": service.get("ip_address"),
                    "domain": service.get("domain"),
                    "externally_detected_providers": service.get(
                        "externally_detected_providers"
                    ),
                    "is_active": service.get("is_active"),
                    "first_observed": service.get("first_observed"),
                    "last_observed": service.get("last_observed"),
                    "port": service.get("port"),
                    "protocol": service.get("protocol"),
                    "active_classifications": service.get("active_classifications"),
                    "inactive_classifications": service.get("inactive_classifications"),
                    "discovery_type": service.get("discovery_type"),
                    "externally_inferred_vulnerability_score": service.get(
                        "externally_inferred_vulnerability_score"
                    ),
                    "externally_inferred_cves": service.get("externally_inferred_cves"),
                    "service_key": service.get("service_key"),
                    "service_key_type": service.get("service_key_type"),
                }

                service_object, created = XpanseServicesMdl.objects.update_or_create(
                    service_id=service.get("service_id"),
                    defaults=service_defaults,
                )

                for cve in service.get("cves", []):
                    cve_defaults = {
                        "cvss_score_v2": cve.get("cvss_score_v2"),
                        "cve_severity_v2": cve.get("cve_severity_v2"),
                        "cvss_score_v3": cve.get("cvss_score_v3"),
                        "cve_severity_v3": cve.get("cve_severity_v3"),
                    }
                    cve_object, created = XpanseCvesMdl.objects.get_or_create(
                        cve_id=cve.get("cve_id"),
                        defaults=cve_defaults,
                    )

                    cve_service_default = {
                        "inferred_cve_match_type": cve.get("inferred_cve_match_type"),
                        "product": cve.get("product"),
                        "confidence": cve.get("confidence"),
                        "vendor": cve.get("vendor"),
                        "version_number": cve.get("version_number"),
                        "activity_status": cve.get("activity_status"),
                        "first_observed": cve.get("first_observed"),
                        "last_observed": cve.get("last_observed"),
                    }

                    (
                        cve_match_object,
                        created,
                    ) = XpanseCveServiceMdl.objects.update_or_create(
                        xpanse_inferred_cve=cve_object,
                        xpanse_service=service_object,
                        defaults=cve_service_default,
                    )

                service_objects.append(service_object)
            alert_object.services.set(service_objects)
            alert_object.save()

        except Exception as e:
            LOGGER.error("Failed to save alert: %s, moving on to the next alert.", e)
            continue
